### ISSUE TITLE
Bug: `no-useless-assignment` false positive in generator functions

### ISSUE THREAD

**Issue by yume-chan** [eslint/eslint#20583]
### Environment

Node version: v22.14.0
npm version: v10.9.2
Local ESLint version: v10.0.2 (Currently used)
Global ESLint version: Not found
Operating System: win32 10.0.26200

### What parser are you using?

Default (Espree)

### What did you do?

<details>
<summary>Configuration</summary>

```js
import js from "@eslint/js";
import globals from "globals";
import { defineConfig } from "eslint/config";

export default defineConfig([
  { files: ["**/*.{js,mjs,cjs}"], plugins: { js }, extends: ["js/recommended"], languageOptions: { globals: globals.browser } },
]);
```
</details>

```js
function* generator() {
  let done = false;
  try {
    yield 1;
    done = true;
  } catch {
    done = true;
  } finally {
    if (!done) {
      console.log('done is false')
    }
  }
}

const iterator = generator()[Symbol.iterator]();
iterator.next();
iterator.return();
```

### What did you expect to happen?

No error, since the `yield` can jump to `finally` directly

### What actually happened?

```text
> pnpm eslint

C:\dev\sandbox\eslint-repro\no-useless-assignment.js
  2:7  error  This assigned value is not used in subsequent statements  no-useless-assignment

✖ 1 problem (1 error, 0 warnings)
```

Running the script prints

```text
> node .\no-useless-assignment.js
done is false
```

Showing the initial value is being used.

### Link to Minimal Reproducible Example

https://github.com/yume-chan/eslint-repro

### Participation

- [ ] I am willing to submit a pull request for this issue.

### AI acknowledgment

- [x] I did not use AI to generate this issue report.
- [ ] (If the above is not checked) I have reviewed the AI-generated content before submitting.

**Comment by DMartens** [comment]
The rule is correct in this case as the initial value of `done` is never read.
It is either set in try or catch and then read in finally.
When removing the initial value of `done`, the rule [does not trigger](https://eslint.org/play/#eyJ0ZXh0IjoiZnVuY3Rpb24qIGdlbmVyYXRvcigpIHtcbiAgbGV0IGRvbmU7XG4gIHRyeSB7XG4gICAgeWllbGQgMTtcbiAgICBkb25lID0gdHJ1ZTtcbiAgfSBjYXRjaCB7XG4gICAgZG9uZSA9IHRydWU7XG4gIH0gZmluYWxseSB7XG4gICAgaWYgKCFkb25lKSB7XG4gICAgICBjb25zb2xlLmxvZygnZG9uZSBpcyBmYWxzZScpXG4gICAgfVxuICB9XG59XG4iLCJvcHRpb25zIjp7InJ1bGVzIjp7Im5vLXVzZWxlc3MtYXNzaWdubWVudCI6WyJlcnJvciJdfSwibGFuZ3VhZ2VPcHRpb25zIjp7InBhcnNlck9wdGlvbnMiOnsiZWNtYUZlYXR1cmVzIjp7ImpzeCI6dHJ1ZX0sInNvdXJjZVR5cGUiOiJtb2R1bGUifSwiZWNtYVZlcnNpb24iOjIwMTUsInBhcnNlciI6IkB0eXBlc2NyaXB0LWVzbGludC9wYXJzZXIifX19).

**Comment by yume-chan** [comment]
Did you read the issue? My repro clear shows `done` in `finally` is still `false` (the initial value)

**Comment by DMartens** [comment]
I am sorry, I misunderstood what you meant.
Yes you are correct that this a false positive, as it seems like the code path analyzer does not take finishing a generator into account while paused.

**Comment by fasttime** [comment]
After the `yield` in this case, execution could resume in the try block or jumps to the catch or finally block, but the code path analysis doesn't seem to account for that:

[**Code Path Graph**](https://explorer.eslint.org/#eslint-explorer=IntcInN0YXRlXCI6e1widG9vbFwiOlwicGF0aFwiLFwiY29kZVwiOntcImphdmFzY3JpcHRcIjpcImZ1bmN0aW9uKiBnZW5lcmF0b3IoKSB7XFxuICB0cnkge1xcbiAgICB5aWVsZCAxO1xcbiAgICBjb25zb2xlLmxvZygnY29udGludWUgaW4gdHJ5IGJsb2NrJyk7XFxuICB9IGNhdGNoIHtcXG4gICAgY29uc29sZS5sb2coJ2NhdGNoIGJsb2NrJyk7XFxuICB9IGZpbmFsbHkge1xcbiAgICBjb25zb2xlLmxvZygnZmluYWxseSBibG9jaycpO1xcbiAgfVxcbn1cXG5cIixcImpzb25cIjpcIi8qKlxcbiAqIFR5cGUgb3IgcGFzdGUgc29tZSBKU09OIGhlcmUgdG8gbGVhcm4gbW9yZSBhYm91dFxcbiAqIHRoZSBzdGF0aWMgYW5hbHlzaXMgdGhhdCBFU0xpbnQgY2FuIGRvIGZvciB5b3UuXFxuICpcXG4gKiBUaGUgdGFicyBhcmU6XFxuICpcXG4gKiAtIEFTVCAtIFRoZSBBYnN0cmFjdCBTeW50YXggVHJlZSBvZiB0aGUgY29kZSwgd2hpY2ggY2FuXFxuICogICBiZSB1c2VmdWwgdG8gdW5kZXJzdGFuZCB0aGUgc3RydWN0dXJlIG9mIHRoZSBjb2RlLiBZb3VcXG4gKiAgIGNhbiB2aWV3IHRoaXMgc3RydWN0dXJlIGFzIEpTT04gb3IgaW4gYSB0cmVlIGZvcm1hdC5cXG4gKlxcbiAqIFlvdSBjYW4gY2hhbmdlIHRoZSB3YXkgdGhhdCB0aGUgSlNPTiBjb2RlIGlzIGludGVycHJldGVkXFxuICogYnkgY2xpY2tpbmcgXFxcIkpTT05cXFwiIGluIHRoZSBoZWFkZXIgYW5kIHNlbGVjdGluZyBkaWZmZXJlbnRcXG4gKiBvcHRpb25zLlxcbiAqXFxuICogVGhpcyBleGFtcGxlIGlzIGluIEpTT05DIG1vZGUsIHdoaWNoIGFsbG93cyBjb21tZW50cy5cXG4gKi9cXG5cXG57XFxuICAgIFxcXCJrZXkxXFxcIjogW3RydWUsIGZhbHNlLCBudWxsXSxcXG4gICAgXFxcImtleTJcXFwiOiB7XFxuICAgICAgICBcXFwia2V5M1xcXCI6IFsxLCAyLCBcXFwiM1xcXCIsIDFlMTAsIDFlLTNdXFxuICAgIH1cXG59XCIsXCJtYXJrZG93blwiOlwiICAgICAgICAgICAgZm9yIChjb25zdCB2YWx1ZSBvZiB2YWx1ZXMpIHtcXG4gICAgICAgICAgICAgICAgcmVzb3VyY2U7IFxcbiAgICAgICAgICAgIH1cIixcImNzc1wiOlwiLyoqXFxuICogVHlwZSBvciBwYXN0ZSBzb21lIENTUyBoZXJlIHRvIGxlYXJuIG1vcmUgYWJvdXRcXG4gKiB0aGUgc3RhdGljIGFuYWx5c2lzIHRoYXQgRVNMaW50IGNhbiBkbyBmb3IgeW91LlxcbiAqXFxuICogVGhlIHRhYnMgYXJlOlxcbiAqXFxuICogLSBBU1QgLSBUaGUgQWJzdHJhY3QgU3ludGF4IFRyZWUgb2YgdGhlIGNvZGUsIHdoaWNoIGNhblxcbiAqICAgYmUgdXNlZnVsIHRvIHVuZGVyc3RhbmQgdGhlIHN0cnVjdHVyZSBvZiB0aGUgY29kZS4gWW91XFxuICogICBjYW4gdmlldyB0aGlzIHN0cnVjdHVyZSBhcyBKU09OIG9yIGluIGEgdHJlZSBmb3JtYXQuXFxuICpcXG4gKiBZb3UgY2FuIGNoYW5nZSB0aGUgd2F5IHRoYXQgdGhlIENTUyBjb2RlIGlzIGludGVycHJldGVkXFxuICogYnkgY2xpY2tpbmcgXFxcIkNTU1xcXCIgaW4gdGhlIGhlYWRlciBhbmQgc2VsZWN0aW5nIGRpZmZlcmVudFxcbiAqIG9wdGlvbnMuXFxuICovXFxuXFxuQGltcG9ydCB1cmwoJ2h0dHBzOi8vZm9udHMuZ29vZ2xlYXBpcy5jb20vY3NzMj9mYW1pbHk9Um9ib3RvOndnaHRANDAwOzcwMCZkaXNwbGF5PXN3YXAnKTtcXG5cXG5ib2R5IHtcXG5cXHRmb250LWZhbWlseTogc2Fucy1zZXJpZjtcXG59XFxuXFxuaDEge1xcblxcdGNvbG9yOiAjMzMzO1xcbn1cXG5cXG5wIHtcXG5cXHRtYXJnaW46IDFlbSAwO1xcbn1cIixcImh0bWxcIjpcIjwhRE9DVFlQRSBodG1sPlxcbjwhLS1cXG5UeXBlIG9yIHBhc3RlIHNvbWUgSFRNTCBoZXJlIHRvIGxlYXJuIG1vcmUgYWJvdXRcXG50aGUgc3RhdGljIGFuYWx5c2lzIHRoYXQgRVNMaW50IGNhbiBkbyBmb3IgeW91LlxcblxcblRoZSB0YWJzIGFyZTpcXG5cXG4tIEFTVCAtIFRoZSBBYnN0cmFjdCBTeW50YXggVHJlZSBvZiB0aGUgY29kZSwgd2hpY2ggY2FuXFxuYmUgdXNlZnVsIHRvIHVuZGVyc3RhbmQgdGhlIHN0cnVjdHVyZSBvZiB0aGUgY29kZS4gWW91XFxuY2FuIHZpZXcgdGhpcyBzdHJ1Y3R1cmUgYXMgSlNPTiBvciBpbiBhIHRyZWUgZm9ybWF0Llxcbi0tPlxcblxcbjxodG1sIGxhbmc9XFxcImVuXFxcIj5cXG4gICAgPGhlYWQ%2BXFxuICAgICAgICA8bWV0YSBjaGFyc2V0PVxcXCJVVEYtOFxcXCI%2BXFxuICAgICAgICA8dGl0bGU%2BSFRNTDwvdGl0bGU%2BXFxuICAgIDwvaGVhZD5cXG4gICAgPGJvZHk%2BXFxuICAgICAgICA8cD5UZXh0PC9wPlxcbiAgICA8L2JvZHk%2BXFxuPC9odG1sPlwifSxcImxhbmd1YWdlXCI6XCJqYXZhc2NyaXB0XCIsXCJqc09wdGlvbnNcIjp7XCJwYXJzZXJcIjpcImVzcHJlZVwiLFwic291cmNlVHlwZVwiOlwic2NyaXB0XCIsXCJlc1ZlcnNpb25cIjpcImxhdGVzdFwiLFwiaXNKU1hcIjp0cnVlfSxcImpzb25PcHRpb25zXCI6e1wianNvbk1vZGVcIjpcImpzb25jXCIsXCJhbGxvd1RyYWlsaW5nQ29tbWFzXCI6ZmFsc2V9LFwiY3NzT3B0aW9uc1wiOntcImNzc01vZGVcIjpcImNzc1wifSxcIm1hcmtkb3duT3B0aW9uc1wiOntcIm1hcmtkb3duTW9kZVwiOlwiY29tbW9ubWFya1wiLFwibWFya2Rvd25Gcm9udG1hdHRlclwiOlwib2ZmXCJ9LFwiaHRtbE9wdGlvbnNcIjp7XCJ0ZW1wbGF0ZUVuZ2luZVN5bnRheFwiOlwibm9uZVwiLFwiZnJvbnRtYXR0ZXJcIjpmYWxzZX0sXCJ3cmFwXCI6dHJ1ZSxcInZpZXdNb2Rlc1wiOntcImFzdFZpZXdcIjpcInRyZWVcIixcInNjb3BlVmlld1wiOlwiZmxhdFwiLFwicGF0aFZpZXdcIjpcImdyYXBoXCJ9LFwicGF0aEluZGV4XCI6e1wiaW5kZXhcIjoxLFwiaW5kZXhlc1wiOjJ9LFwiZXNxdWVyeVNlbGVjdG9yXCI6e1wic2VsZWN0b3JcIjpcIlwifX0sXCJ2ZXJzaW9uXCI6MH0i)

Here's a repro that shows all three paths:

```js
function* generator() {
  try {
    yield 1;
    console.log('continue in try block');
  } catch {
    console.log('catch block');
  } finally {
    console.log('finally block');
  }
}

console.log('iterator.next() :');
const iterator1 = generator()[Symbol.iterator]();
iterator1.next();
iterator1.next();

console.log('\niterator.return() :');
const iterator2 = generator()[Symbol.iterator]();
iterator2.next();
iterator2.return();

console.log('\niterator.throw() :');
const iterator3 = generator()[Symbol.iterator]();
iterator3.next();
iterator3.throw(new Error);
```

**Comment by sethamus** [comment]
I think this is a false positive as well. I can work on this.

**Comment by techmannih** [comment]
May I work on in issue

**Comment by meganprice661-svg** [comment]
Do so and update me later on

**Comment by snitin315** [comment]
Marking this as accepted.

**Comment by Kuldeep2822k** [comment]
can i procced to solve this issue or is anyone working on it

**Comment by meganprice661-svg** [comment]
Thanks for the update
